Add support for opening files for appending

zsol opened the pull request:
This is a reimplementation of #8984 based on the end result of the discussion over there.

Addresses part of #8983

davide125 commented:
The Rawhide CI failure here looks unrelated

poettering commented:
patch looks pretty OK otherwise.

(note though that we usually prefer merging "perfect" PRs, i.e. the commits that make up a PR should reflect logical steps, not historical ones, hence please squash any commits that just fix mistakes of earlier ones in the same PR)

zsol commented:
Thanks for the feedback, I'll iterate on this soon (truncate -> create with the tempfile semantics, hopefully avoiding the name).

> please squash any commits that just fix mistakes

I'm happy to do this if you prefer it like that. On my other open source projects I just let github do the squashing, because I find reviewing small incremental changes to the PR easier.

keszybz commented:
> I just let github do the squashing

The PR should be composed of as many logical patches as needed, each one doing a separate thing. Github can only squash everything into one patch, but if more than one patch is wanted, it cannot help with that.

keszybz commented:
… and even if there's just one patch, the "squash & merge" option sometimes mangles patch authorship. So it *is* overall better if you do the squashing yourself.

zsol commented:
I reworked truncation (and renamed it to `create:` because it seemed more accurate) and fixed all the tests so they actually fail if something breaks :)

Unfortunately I couldn't use `O_TMPFILE` as the only way to give it a name is with `linkat`, which won't overwrite existing files. I went with `mkostemp` and `renameat` instead, but unfortunately this doesn't seem to work either. The file is properly created, but I can't seem to get any output in it. I'd appreciate some help here :)

zsol commented:
Thanks for the help and thorough review, I don't think I would have gotten to the bottom of it alone.

> consider reusing the same fd (it would have to be stored in systemd for the duration of the service), in case Type=oneshot and multiple commands ExecStart= commands are specified. This would require the fd to be stored in pid1 for the duration of the service. But it would match what happens with SI=socket more closely.

This sounds like the right approach to me, and I'd be happy to take a crack at it as a separate PR as it seems to be a bigger change. So how about this: let me drop support for `create:` in this PR, as `append:` and `file:` still works fine. I'll address the review comments, and then we can merge this. Then I'll open a new pull request to both make the above happen and add `create:`.

zsol commented:
Looks like the tests fail if I add `PrivateTmp=true`, but pass if I remove it. I wonder if there's a similar problem with setting up the private mount namespace per ES command?

poettering commented on man/systemd.exec.xml:
```diff
@@ -1749,8 +1749,8 @@ SystemCallErrorNumber=EPERM</programlisting>
-        <option>file:<replaceable>path</replaceable></option>, <option>socket</option> or
-        <option>fd:<replaceable>name</replaceable></option>.</para>
+        <option>file:<replaceable>path</replaceable></option>, <option>append:<replaceable>path</replaceable></option>,
+        , <option>socket</option> or<option>fd:<replaceable>name</replaceable></option>.</para>
```
one "," at the end of the line, and another one at the beginning of the next?

poettering commented on src/core/execute.c:
```diff
@@ -675,9 +676,11 @@ static int setup_output(
+        case EXEC_OUTPUT_FILE_APPEND: {
                 bool rw;
                 int fd;
+                int flags;
```
not that it matters, but why not merge this with the previous line...

zsol commented:
Fair enough! Fixed the typos and changed the commit message to not close #8983 just reference it.

poettering commented:
This breaks the ubuntu CI tests:

```
Child 9500 (cat) died (code=exited, status=0/SUCCESS)
exec-standardoutput-file.service: Child 9500 belongs to exec-standardoutput-file.service.
exec-standardoutput-file.service: Main process exited, code=exited, status=0/SUCCESS
exec-standardoutput-file.service: Running next main command for state start.
exec-standardoutput-file.service: Passing 0 fds to service
exec-standardoutput-file.service: About to execute: /usr/bin/cmp /tmp/test-exec-standardoutput-expected /tmp/test-exec-standardoutput-output
exec-standardoutput-file.service: Forked /usr/bin/cmp as 9501
Received SIGCHLD from PID 9501 (cmp).
Child 9501 (cmp) died (code=exited, status=1/FAILURE)
exec-standardoutput-file.service: Child 9501 belongs to exec-standardoutput-file.service.
exec-standardoutput-file.service: Main process exited, code=exited, status=1/FAILURE
exec-standardoutput-file.service: Failed with result 'exit-code'.
exec-standardoutput-file.service: Changed start -> failed
exec-standardoutput-file.service: Unit entered failed state.
Assertion 'service->main_exec_status.status == status_expected' failed at ../src/test/test-execute.c:59, function check(). Aborting.
Aborted (core dumped)
FAIL: test-execute (code: 134)
```

zsol commented:
Yep I'll take a look

On Sat, Jul 14, 2018, 12:57 Lennart Poettering <notifications@github.com>
wrote:

> This breaks the ubuntu CI tests:
>
> Child 9500 (cat) died (code=exited, status=0/SUCCESS)
> exec-standardoutput-file.service: Child 9500 belongs to exec-standardoutput-file.service.
> exec-standardoutput-file.service: Main process exited, code=exited, status=0/SUCCESS
> exec-standardoutput-file.service: Running next main command for state start.
> exec-standardoutput-file.service: Passing 0 fds to service
> exec-standardoutput-file.service: About to execute: /usr/bin/cmp /tmp/test-exec-standardoutput-expected /tmp/test-exec-standardoutput-output
> exec-standardoutput-file.service: Forked /usr/bin/cmp as 9501
> Received SIGCHLD from PID 9501 (cmp).
> Child 9501 (cmp) died (code=exited, status=1/FAILURE)
> exec-standardoutput-file.service: Child 9501 belongs to exec-standardoutput-file.service.
> exec-standardoutput-file.service: Main process exited, code=exited, status=1/FAILURE
> exec-standardoutput-file.service: Failed with result 'exit-code'.
> exec-standardoutput-file.service: Changed start -> failed
> exec-standardoutput-file.service: Unit entered failed state.
> Assertion 'service->main_exec_status.status == status_expected' failed at ../src/test/test-execute.c:59, function check(). Aborting.
> Aborted (core dumped)
> FAIL: test-execute (code: 134)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/systemd/systemd/pull/9500#issuecomment-405018441>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAEEtKxETkFYpNlsCzo5_P_oHQlLnbPzks5uGdyxgaJpZM4VBguu>
> .
>

zsol commented:
@poettering where did you see that log? I can't find it by clicking on the bionic-* checks's details

poettering commented:
@zsol hmm, so the CI was borked for unrelated reasons on your repush. See #9621.

The old CI showed the issue... Can you rebase on current git and repush? Hopefully the ubuntu CI gets far enough then, as #9621 has been fixed.

zsol commented:
Okay I realize what's happening here. ubuntu has a more minimal version of `/bin/sh` where the builtin `echo` doesn't recognize `-e`:

```
output(14): hello
o
hello
x
expected(14): -e hello
ello
x
Assertion 'streq(output, expected)' failed at ../src/test/test-execute.c:635, function test_exec_standardoutput(). Aborting.
```

poettering commented:
Oh yeah, they have "dash", a major source of unnecessary bugs... I figure you can just use /bin/echo instead.

zsol commented:
Yes, that would've been good, too. I went with `printf` :)

filbranden commented on test/test-execute/exec-standardoutput-append.service:
```diff
@@ -0,0 +1,13 @@
+Description=Test for StandardOutput=append:
+
+[Service]
+ExecStartPre=sh -c 'printf hello\\\\n > /tmp/test-exec-standardoutput-output'
```
You can also let systemd expand the newlines on itself... If you use a bare \n here it will do that. As long as the string looks quoted for the shell itself, you'll be fine.

See also an example here: https://github.com/systemd/systemd/blob/v239/test/TEST-04-JOURNAL/test.sh#L38

This should work I think:

```
ExecStartPre=sh -c 'printf "hello\n" > /tmp/test-exec-standardoutput-output'
ExecStartPre=sh -c 'printf "hello\nhello\n" > /tmp/test-exec-standardoutput-expected'
```

Or just use `echo` to avoid one of the `\n`s:

```
ExecStartPre=sh -c 'echo "hello" > /tmp/test-exec-standardoutput-output'
ExecStartPre=sh -c 'echo "hello\nhello" > /tmp/test-exec-standardoutput-expected'
```

(Another option is to use command grouping on the second line, to run two separate "echo" commands and still keep a single redirection):

```
ExecStartPre=sh -c 'echo "hello" > /tmp/test-exec-standardoutput-output'
ExecStartPre=sh -c '{ echo "hello"; echo "hello"; } > /tmp/test-exec-standardoutput-expected'
```

I think I like the middle one most... It's a bit annoying that the `\n` is expanded by systemd while the rest is shell, but oh well that's systemd unit parsing for you :-)

Cheers,
Filipe

zsol commented:
These new CI errors look unrelated :(

poettering commented:
CI passed now. Let's merge!

zsol commented:
whew, what a journey. Thanks everyone for all the help. I'll be opening the next one about `truncate:` soon
